library/core/src/cmp.rs

```diff
@@ -274,6 +274,7 @@ pub trait Eq: PartialEq<Self> {
     //
     // This should never be implemented by hand.
     #[doc(hidden)]
+    #[cfg_attr(not(bootstrap), no_coverage)]
```

Comment:
Should `#[no_coverage]` be behind a feature gate? I think new attributes need to go through the lang team.

Reply:
@jyn514 I don't think I have permission to mention a group like @rust-lang/lang ... If this doesn't add them, can you do it?

I'm fairly sure `fn assert_receiver_is_total_eq()` is some kind of marker function that is never meant to be called. We need some way to mark a function like this as "hide from coverage instrumentation", so to speak. A function attribute makes the most sense to me, and I tried to align the naming and implementation to how similar attributes are handled.

If a feature gate is also required, let me know.

Reply:
Right, yes, adding a perma-unstable attribute is fine. But adding a new attribute that *any* user of the compiler can use needs approval from the language team.

Another alternative is to special-case functions with an empty body to not need coverage. But the attribute seems simpler.

Reply:
We do want to instrument most functions with an empty body, actually.

Please add the language team to this PR (since I apparently still can't do it myself). Thanks!

Reply:
cc @rust-lang/lang

Reply:
Yup, defining a meaning for an attribute needs an FCP from an appropriate team as the attribute would have to be there forever to avoid breaking code.  (Right now it's "error: cannot find attribute `no_coverage` in this scope".)

Could this maybe be a tool attribute, somehow?  Having it be `#[name_of_tool::no_coverage]` lowers the bar somewhat, AFAIK.

Reply:
It's handled by rustc's `-Z instrument-coverage` option (which should eventually become a stablized compiler option).

Reply:
Current tracking issue: #79121

This also falls under the MCP for coverage: rust-lang/compiler-team#278

Reply:
We can put this behind a feature gate for now to get this fix in. ([docs on this](https://rustc-dev-guide.rust-lang.org/feature-gates.html)). I suspect this will be a useful attribute to have in the language, so I don't think it should be perma-unstable.

You could argue for putting this in a tool namespace, but it's really a compiler flag with some associated LLVM tooling, so there's not any precedent for that as far as I'm aware. The name that comes to mind is `profiling` since it's broad enough to cover similar uses of compiler instrumentation.

Reply:
I've been trying to set up the feature gate, but the function attr is applied to the function derived from the `Eq` trait. Since I'm only enabling the feature in the core `cmp` crate, the compiler complains on all struct declarations that derive `Eq`:

```
error[E0658]: the `#[no_coverage]` attribute is an experimental feature
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/cpuid.rs:10:30
    |
10  |   #[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
    |                                ^^ in this macro invocation
    | 
   ::: library/core/src/cmp.rs:288:1
    |
288 | / pub macro Eq($item:item) {
289 | |     /* compiler built-in */
290 | | }
    | |_- in this expansion of `#[derive(Eq)]`
    |
    = note: see issue #84605 <https://github.com/rust-lang/rust/issues/84605> for more information
    = help: add `#![feature(no_coverage)]` to the crate attributes to enable
```

Reply:
It might be easier to special case this function in the compiler - see if it's named `assert_receiver_is_total_eq` and it's implementing the `Eq` trait defined in core. Then you don't have to worry about the new attribute.

Reply:
I would rather not implement a special case. Since we're aiming to stablize the `instrument-coverage` compiler option, we should do this the right way.

I realized the `no_coverage` attribute is added to the derived `Eq` impl via:

https://github.com/rust-lang/rust/blob/e977835012ea5196a26160304ad90c7d7f21b61c/compiler/rustc_builtin_macros/src/deriving/cmp/eq.rs

I had added the `no_coverage` attribute to the `MethodDef`, and thought maybe I also needed to add the `feature(no_coverage)` attribute in the same way. That doesn't appear to be working though.

The current change (even though it's not working) is uploaded to this PR.

Any thoughts on whether this should work, to enable a gated feature via a `MethodDef` attribute? Or if not, how then?

Thanks

Reply:
I'm in favor of the feature gate. I think we probably do want to stabilize the attribute immediately.

Reply:
Sounds good. Thanks Niko.

I think I understand what I'm doing wrong and how to make the feature gate work. I assume the `feature` is just another attribute, and I can probably change the gating to check for the `feature(no_coverage)` attribute at either the function or crate level. This should allow me to enable the feature _only_ on the `assert_receiver_is_total_eq()` function, without adding the feature in every crate that derives `Eq` (which would generally make the gate not a gate at all). 

I think I can figure that out, assuming this is a fairly unique situation.

If there are other examples of function-level feature gating, please let me know.

Reply:
I got the gating to work! (supports either individual function enabling or crate-level). Tests added to confirm gating is checked, and also that gating works (crate and function).

The `Eq` coverage issue is addressed by this PR as well, of course.